modulus/models/afno/modafno.py

```diff
@@ -0,0 +1,649 @@
+            sparsity_threshold=sparsity_threshold,
+            hard_thresholding_fraction=hard_thresholding_fraction,
+            hidden_size_factor=hidden_size_factor,
+        )
+
+        self.scale_shift_mode = scale_shift_mode
+        self.channel_mul = 1 if scale_shift_mode == "real" else 2
+        if scale_shift_kwargs is None:
+            scale_shift_kwargs = {}
+        self.scale_shift = ScaleShiftMlp(
+            mod_features,
+            self.num_blocks
+            * self.block_size
+            * self.hidden_size_factor
+            * self.channel_mul,
+            **scale_shift_kwargs,
+        )
+
+    def forward(self, x: Tensor, mod_embed: Tensor) -> Tensor:
+        bias = x
+
+        dtype = x.dtype
+        x = x.float()
+        B, H, W, C = x.shape
+        # Using ONNX friendly FFT functions
+        x = fft.rfft2(x, dim=(1, 2), norm="ortho")
+        x_real, x_imag = fft.real(x), fft.imag(x)
+        x_real = x_real.reshape(B, H, W // 2 + 1, self.num_blocks, self.block_size)
+        x_imag = x_imag.reshape(B, H, W // 2 + 1, self.num_blocks, self.block_size)
+        o1_shape = (
+            B,
+            H,
+            W // 2 + 1,
+            self.num_blocks,
+            self.block_size * self.hidden_size_factor,
+        )
+        scale_shift_shape = (B, self.channel_mul, 1, o1_shape[3], o1_shape[4])
+
+        o1_real = torch.zeros(o1_shape, device=x.device)
+        o1_imag = torch.zeros(o1_shape, device=x.device)
+        o2 = torch.zeros(x_real.shape + (2,), device=x.device)
+
+        total_modes = min(H, W) // 2 + 1
+        kept_modes = int(total_modes * self.hard_thresholding_fraction)
+
+        o1_re = (
+            torch.einsum(
+                "nyxbi,bio->nyxbo",
+                x_real[
+                    :, total_modes - kept_modes : total_modes + kept_modes, :kept_modes
+                ],
+                self.w1[0],
+            )
+            - torch.einsum(
+                "nyxbi,bio->nyxbo",
+                x_imag[
+                    :, total_modes - kept_modes : total_modes + kept_modes, :kept_modes
+                ],
+                self.w1[1],
+            )
+            + self.b1[0]
+        )
+
+        o1_im = (
+            torch.einsum(
+                "nyxbi,bio->nyxbo",
+                x_imag[
+                    :, total_modes - kept_modes : total_modes + kept_modes, :kept_modes
+                ],
+                self.w1[0],
+            )
+            + torch.einsum(
+                "nyxbi,bio->nyxbo",
+                x_real[
+                    :, total_modes - kept_modes : total_modes + kept_modes, :kept_modes
+                ],
+                self.w1[1],
+            )
+            + self.b1[1]
+        )
+
+        # scale-shift operation
+        (scale, shift) = self.scale_shift(mod_embed)
+        scale = scale.view(*scale_shift_shape)
+        shift = shift.view(*scale_shift_shape)
+        if self.scale_shift_mode == "real":
```

Comment:
In the constructor there is presently no check if `scale_shift_mode` is real or complex.

I would suggest either line 178 gets changed to through an error if not one of these, or line 260 gets changed to an else for consistency.

Reply:
Added error check to constructor in https://github.com/NVIDIA/modulus/pull/547/commits/4750b4d689943825b371709428cd1ae51289de59.